sw/host/opentitanlib/src/test_utils/bitbanging/i2c.rs

```diff
@@ -0,0 +1,347 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#[repr(u8)]
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Bit {
+    Low = 0,
+    High = 1,
+}
+
+impl std::convert::From<u8> for Bit {
+    fn from(val: u8) -> Self {
+        match val {
+            0x00 => Self::Low,
+            0x01 => Self::High,
+            _ => panic!("Can't convert"),
+        }
+    }
+}
+pub mod encoder {
+    use super::*;
+
+    #[derive(Debug, PartialEq)]
+    pub enum Transfer<'w> {
+        Start,
+        Stop,
+        Addr { addr: u8, read: bool, nack: bool },
+        Write(&'w [u8]),
+        Read(usize),
+        Broken(Vec<Bit>),
+    }
+
+    impl Transfer<'_> {
+        fn bitbaging<const SDA: u8, const SCL: u8>(&self, is_next_stop: bool) -> Vec<u8> {
```

Comment:
nit:
```suggestion
        fn bitbanging<const SDA: u8, const SCL: u8>(&self, is_next_stop: bool) -> Vec<u8> {
```

Maybe to avoid intermediate Vec creations?
```suggestion
        fn bitbaging<const SDA: u8, const SCL: u8>(&self, is_next_stop: bool, samples: &mut Vec<u8>) {
```

this could allow use of `Symbol` abstraction without additional overhead

If we have
```rust
impl Symbol {
    fn bitbanging<const SDA: u8, const SCL: u8>(&self, samples: &mut Vec<u8>) {
        match self {
            Symbol::Start => samples.extend([...]),
            Symbol::Byte(...) => /* content by fn byte */
            Symbol::Broken(...) => /* content by fn bits */
            ...
        }
    }
}
```
then the function can be simplified as
```rust
match self {
    Transfer::Start => Symbol::Start.bitbanging::<SDA, SCL>(samples),
    Transfer::Stop => Symbol::Start.bitbanging::<SDA, SCL>(samples),
    Transfer::Addr { addr, read, nack } => (Symbol::Byte { data: addr << 1 | *read as u8, nack }).bitbanging::<SDA, SCL>(samples),
    ...
}
```